docs/proposals/deploy.md

```diff
@@ -0,0 +1,152 @@
+- **Pause/resume**: To pause a deployment mid-way, and to resume it. (A use case is to support canary deployment.)
+- **Version information**: To record and show version information that's meaningful to users. This can be useful for rollback.
+
+## Related `kubectl` Commands
+
+### `kubectl run`
+
+`kubectl run` should support the creation of Deployment (already implemented) and DaemonSet resources.
+
+### `kubectl scale` and `kubectl autoscale`
+
+Users may use `kubectl scale` or `kubectl autoscale` to scale up and down Deployments (both already implemented).
+
+### `kubectl rollout`
+
+`kubectl rollout` supports both Deployment and DaemonSet. It has the following subcommands:
+- `kubectl rollout undo` works like rollback; it allows the users to rollback to a previous version of deployment.
+- `kubectl rollout pause` allows the users to pause a deployment.
+- `kubectl rollout resume` allows the users to resume a paused deployment.
+- `kubectl rollout status` shows the status of a deployment.
+- `kubectl rollout history` shows meaningful version information of all previous deployments.
+
+### `kubectl set`
+
+`kubectl set` has the following subcommands:
+- `kubectl set env` allows the users to set environment variables of Kubernetes resources. It should support Pod, ReplicationController, ReplicaSet, Deployment, and DaemonSet.
+- `kubectl set image` allows the users to update the image of deployments. Users will use `--container` and `--image` flags to update the image of a container.
+
+### Example
+
+With the commands introduced above, here's an example of deployment management:
+
+```console
+# Create a Deployment
+$ kubectl run nginx --image=nginx --replicas=2 --generator=deployment/v1beta1
+
+# Watch the Deployment status
+$ kubectl rollout status deployment/nginx
+
+# Update the Deployment 
+$ kubectl set image deployment/nginx --container=nginx --image=nginx:<some-version>
+
+# Pause the Deployment
+$ kubectl rollout pause deployment/nginx
+
+# Resume the Deployment
+$ kubectl rollout resume deployment/nginx
+
+# Check the change history (deployment versions)
+$ kubectl rollout history deployment/nginx
+
+# Rollback to a previous version.
+$ kubectl rollout undo deployment/nginx --to-version=<version>
+```
+
+## Support in Deployment
+
+### Deployment Status
+
+Deployment status should summarize information about Pods, which includes:
+- The number of pods of each version.
+- The number of ready/not ready pods.
+
+See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
+
+### Deployment Version
+
+We store previous deployment versions information in deployment annotation `kubectl.kubernetes.io/deployment-version-<time>`. The value stored in the annotation can be a spec hash, a strategic merge patch or the kubectl commands previously executed. We choose strategic merge patch since it's more human-readable than spec hash, and commands like `kubectl edit` won't provide enough useful information.
```

Comment:
@janetkuo It would help to explain what this would be used for. For instance, would it be used to support rolling back changes? Or just to view changes? Note also that we could support multiple annotations. For instance, for viewing history, I think recording the source of the mutation, such as the kubectl command (`rollout.kubectl.kubernetes.io/change-source="kubectl set image deployment/nginx --image=nginx:1.9.8"`), would be useful.

Also, I think it would help to clarify whether this annotation belongs on Deployment or on the underlying ReplicaSets that are generated by mutations of Deployment, or both. If on Deployment, should there be more than one?